Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -174,42 +178,47 @@ class TopTabsViewController: UIViewController {
     }
     
     func newTabTapped() {
-        if let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab),
-            let cell  = collectionView.cellForItemAtIndexPath(NSIndexPath(forItem: index, inSection: 0)) as? TopTabCell {
-            cell.selectedTab = false
-            if index > 0 {
-                cell.seperatorLine = true
-            }
+        dispatch_async(dispatch_get_main_queue()) {
+            self.delegate?.topTabsDidPressNewTab(self.isPrivate)
         }
-        delegate?.topTabsDidPressNewTab()
-        self.privateModeButton.enabled = false
-        collectionView.performBatchUpdates({ _ in
-            let count = self.collectionView.numberOfItemsInSection(0)
-            self.collectionView.insertItemsAtIndexPaths([NSIndexPath(forItem: count, inSection: 0)])
-            }, completion: { finished in
-                if finished {
-                    self.privateModeButton.enabled = true
-                    self.scrollToCurrentTab()
-                }
-        })
     }
-    
+
     func togglePrivateModeTapped() {
-        delegate?.topTabsDidPressPrivateModeButton(isPrivate ? lastNormalTab : lastPrivateTab)
+        if isUpdating || pendingReloadData {
+            return
+        }
+        delegate?.topTabsDidPressPrivateModeButton()
+        self.pendingReloadData = true
+        let oldSelectedTab = self._oldSelectedTab
+        self._oldSelectedTab = tabManager.selectedTab
         self.privateModeButton.setSelected(isPrivate, animated: true)
-        self.collectionView.reloadData()
-        self.scrollToCurrentTab(false, centerCell: true)
-    }
-    
-    func closeTab() {
-        delegate?.topTabsDidPressTabs()
+
+        //if private tabs is empty and we are transitioning to it add a tab
+        if tabManager.privateTabs.isEmpty  && !isPrivate {
+            tabManager.addTab(isPrivate: true)
+        }
+
+        //get the tabs from which we will select which one to nominate for tribute (selection)
+        //the isPrivate boolean still hasnt been flipped. (It'll be flipped in the BVC didSelectedTabChange method)
+        let tabs = !isPrivate ? tabManager.privateTabs : tabManager.normalTabs
+        if let tab = oldSelectedTab where tabs.indexOf(tab) != nil {
+            tabManager.selectTab(tab)
+        } else {
+            tabManager.selectTab(tabs.last)
+        }
     }
-    
-    func reloadFavicons() {
-        self.collectionView.reloadData()
+
+    func reloadFavicons(notification: NSNotification) {
+        dispatch_async(dispatch_get_main_queue()) {
```

Comment:
Hmm this one is from an NSNotification which can be called from a different thread. Usually notifications happen on the same thread as the ones they were initiated from. So made sense here to use the main queue. Don't want animations happening outside of the main thread.